python/cudf/cudf/core/udf/function.cu

```diff
@@ -0,0 +1,555 @@
+
+  int tid     = threadIdx.x;
+  int tb_size = blockDim.x;
+
+  // Calculate how many elements each thread is working on
+  T local_max            = init_val;
+  int64_t local_idx           = -1;
+
+  __shared__ T smax;
+  __shared__ int64_t sidx;
+
+  if (tid == 0) {
+    smax = init_val;
+    sidx = INT64_MAX;
+  }
+
+  __syncthreads();
+
+// Calculate local max for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (tid + (item * tb_size) < size) {
+      T load = data[tid + item * tb_size];
+      if (load > local_max) {
+        local_max = load;
+        local_idx = index[tid + item * tb_size];
+      }
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local max for each group
+  atomicMax((&smax), local_max);
+
+  __syncthreads();
+
+  if (local_max == smax) { atomicMin((&sidx),local_idx); }
+
+  __syncthreads();
+
+  return sidx;
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ T device_idxmin(T const* data, int const items_per_thread, int64_t const* index, size_type size, T init_val) {
+
+  int tid = threadIdx.x;
+  int tb_size = blockDim.x;
+
+  T local_min            = init_val;
+  int64_t local_idx           = -1;
+
+  __shared__ T smin;
+  __shared__ int64_t sidx;
+
+  if (tid == 0) {
+    smin = init_val;
+    sidx = INT64_MAX;
+  }
+
+  __syncthreads();
+
+// Calculate local max for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (tid + (item * tb_size) < size) {
+      T load = data[tid + item * tb_size];
+      if (load < local_min) {
+        local_min = load;
+        local_idx = index[tid + item * tb_size];
+      }
+    }
+  }
+
+  __syncthreads();
+
+  // Calculate local max for each group
+  atomicMin((&smin), local_min);
+
+  __syncthreads();
+
+  if (local_min == smin) { atomicMin((&sidx), local_idx); }
+
+  __syncthreads();
+
+  return sidx;
+}
+
+extern "C" __device__ int BlockSum_int64(int64_t* numba_return_value,
+                                         int64_t const* data,
+                                         int64_t size)
+{
+  int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  auto const items_per_thread = (size + tb_size - 1) / tb_size;
+  
+  int64_t sum = device_sum<int64_t>(data, items_per_thread, size);
+
+  *numba_return_value = sum;
+
+  return 0;
+}
+
+extern "C" __device__ int BlockSum_float64(double* numba_return_value,
+                                           double const* data,
+                                           int64_t size)
+{
+  int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  auto const items_per_thread = (size + tb_size - 1) / tb_size;
+
+  double sum = device_sum<double>(data, items_per_thread, size);
+
+  *numba_return_value = sum;
+
+  return 0;
+}
```

Comment:
Can we produce less copy-paste by judicious application of templating + preprocessing (not sure if the latter is considered kosher in libcudf): https://gcc.godbolt.org/z/Mah8hG8YP

Reply:
Done